### PR TITLE
[CLIPPER-132] Updating Clipper tutorial

### PR DESCRIPTION
**Pull request by nishadsingh1**
The edits are described in the JIRA Issue: https://clipper.atlassian.net/browse/CLIPPER-132

More than happy to revert some of the additions made here – a lot of them were made because of personal preference. Let me know what you think and what you'd change.

### CODE REVIEW

**Comment by dcrankshaw** [comment]
jenkins, test this please

**Comment by dcrankshaw** [comment]
Jenkins test this please

File: examples/tutorial/tutorial_part_one.ipynb [hunk 1]
```diff
@@ -29,17 +30,46 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, we must extract the data into a format we can load. This will make use of the provided [`extract_cifar.py`](extract_cifar.py)\n",
+    "\n",
+    "This dataset has 50,000 training datapoints and 10,000 test datapoints. We don't need to use all of them to demonstrate how Clipper works. Feel free to adjust `max_train_datapoints` and `max_test_datapoints` or set them to `None` if you wish to use all the data available for training and testing. You can change these vaues and re-run this command in the future if you wish."
+   ]
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_one.ipynb]: In the JIRA issue, you said "We default to using 10,000 training datapoints (instead of the full 50,000). With 50,000 datapionts, writing to the CSV takes a long time, as does training on all the datapoint later. Using 10000 for training yields similar accuracies and prevents some of the existing delays in the tutorial runthrough."

Can you add an aside (using the quote formatting in markdown) that explains this?

File: examples/tutorial/tutorial_part_one.ipynb [hunk 2]
```diff
@@ -91,16 +121,20 @@
     "\n",
     "Now you're ready to start Clipper! You will be using the `clipper_manager` client library to perform admninistrative commands.\n",
     "\n",
-    "If you'd like to deploy Clipper locally, you can leave the `user` and `key` variables blank and set `host=\"local\"`. Otherwise, you can deploy Clipper remotely to a machine that you have SSH access to. Set the `user` variable to your SSH username, the `key` variable to the path to your SSH key, and the `host` variable to the remote hostname or IP address.\n",
+    "> *Remember, Docker and Docker-Compose must be installed to before deploying Clipper. Visit https://docs.docker.com/compose/install/ for instructions on how to do so.*\n",
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_one.ipynb]: Typo: s/be installed to before/be installed before

File: examples/tutorial/tutorial_part_one.ipynb [hunk 3]
```diff
@@ -91,16 +121,20 @@
...
+    "\n",
+    "Clipper uses Docker to manage application configurations and to deploy machine-learning models. Make sure your Docker daemon, local or remote, is up and running. You can check this by running `docker ps` in your command line – if your Docker daemon is not running, you will be told explicitly.\n",
+    "\n",
+    "Starting clipper will have the following effect on your setup:  <img src=\"img/start_clipper.png\" style=\"width: 500px;\"/>\n",
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_one.ipynb]: Clipper should be uppercase (it's a proper noun).

File: examples/tutorial/tutorial_part_one.ipynb [hunk 4]
```diff
@@ -226,7 +262,9 @@
     "http://HOSTNAME:1337/cifar_demo/update\n",
     "```\n",
     "\n",
-    "You will now start querying Clipper with a simple Python frontend app that computes the average accuracy of the responses after every 100 requests and updates a plot of the results with every iteration."
+    "You will now start querying Clipper with a simple Python frontend app that computes the average accuracy of the responses after every 100 requests and updates a plot of the results with every iteration.\n",
+    "\n",
+    "This diagram shows how how the accuracy plot is receiving its test predictions: <img src=\"img/serve_predictions.png\" style=\"width: 500px;\"/>"
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_one.ipynb]: Typo: Remove the extra "how"

File: examples/tutorial/tutorial_part_two.ipynb [hunk 5]
```diff
@@ -33,7 +33,26 @@
     "key = \"\"\n",
     "# Set the SSH host\n",
     "host = \"\"\n",
-    "clipper = cm.Clipper(host, user, key)"
+    "clipper = cm.Clipper(host, user, key)\n",
+    "clipper.start()"
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: Because we're moving cleanup to the end of part 2, there's no reason that clipper should need to be started again. Delete this line.

File: examples/tutorial/tutorial_part_two.ipynb [hunk 6]
```diff
@@ -33,7 +33,26 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Make sure that `cifar_demo` is still registered with Clipper. If it isn't, go back to part one of the tutorial and re-register the app."
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: Delete this cell

File: examples/tutorial/tutorial_part_two.ipynb [hunk 7]
```diff
@@ -33,7 +33,26 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "clipper.list_apps()"
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: Delete this cell

File: examples/tutorial/tutorial_part_two.ipynb [hunk 8]
```diff
@@ -69,7 +88,7 @@
    "source": [
     "# Train Logistic Regression Model\n",
     "\n",
-    "When tackling a new problem with machine learning, it's always good to start with simple models and only add complexity when needed. Start by training a logistic regression binary classifier using [Scikit-Learn](http://scikit-learn.org/). This model gets about 74% accuracy on the offline evaluation dataset."
+    "When tackling a new problem with machine learning, it's always good to start with simple models and only add complexity when needed. Start by training a logistic regression binary classifier using [Scikit-Learn](http://scikit-learn.org/). This model gets about 68% accuracy on the offline evaluation dataset if you use 10000 training datapoints about 74% if you use all training datapoints."
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: Run on sentence. Change it to:
"This model gets about 68% accuracy on the offline evaluation dataset if you use 10000 training examples. It gets about 74% if you use all 50,000 examples."

File: examples/tutorial/tutorial_part_two.ipynb [hunk 9]
```diff
@@ -94,12 +113,15 @@
    "source": [
     "# Deploy Logistic Regression Model\n",
     "\n",
-    "While 74% accuracy on a CIFAR binary classification task is significantly below state of the art, it's already much better than the 50% accuracy of random guessing that you're getting right now without machine learning.\n",
+    "While 68-74% accuracy on a CIFAR binary classification task is significantly below state of the art, it's already much better than the 50% accuracy of random guessing that you're getting right now without machine learning.\n",
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: Change "that you're getting" to "that your application is getting." Also, if you can think of a better verb than "getting" go ahead and change it (I know that was my word choice).

File: examples/tutorial/tutorial_part_two.ipynb [hunk 10]
```diff
@@ -94,12 +113,15 @@
...
     "\n",
     "You can deploy your logistic regression model directly to Clipper without having to worry about how to serialize the model or integrate it with application code.\n",
     "\n",
     "To deploy a model to Clipper, you must assign it a name (\"sklearn_cifar\"), a version (1), and then provide some metadata about the model itself. In this case, you are specifying that you want to run the model using the `sklearn_cifar_container` Docker image in the [Clipper repo](https://hub.docker.com/u/clipper/dashboard/) on Docker Hub. You can assign the model descriptive labels, and specify the input type that this model expects. Finally, you can specify how many *replicas* of the model (how many Docker containers) to launch. Adding more replicas increases the throughput of this model.\n",
     "\n",
+    "After completing this step, Clipper will be managing a container in Docker with your model in it:\n",
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: s/managing a container/managing a new container

File: examples/tutorial/tutorial_part_two.ipynb [hunk 11]
```diff
@@ -209,6 +234,9 @@
     "\n",
     "When you register an application with multiple candidate models, the selection policy dictates which of those models to query, and if more than one model is queried, how to combine the results. Selection policies can be used to implement ensembles, to perform dynamic online model selection using multi-armed bandit algorithms (a more statistically efficient alternative to A/B testing), to estimate prediction confidence, and many other applications. Selection policies can learn from feedback and be updated in real-time directly in the serving system. For more information about selection policies in Clipper, check out the [research paper on Arxiv](https://arxiv.org/abs/1612.03079).\n",
     "\n",
+    "This diagram illustrates how the selection policy affects prediction serving with your current setup:\n",
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: "This diagram illustrates how the EXP4 selection policy sends each prediction query to _all_ the deployed models and then combines their predictions into a single prediction that is returned to the client:"

File: examples/tutorial/tutorial_part_two.ipynb [hunk 12]
```diff
@@ -232,7 +260,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "If you check out the accuracy of your frontend application again in the notebook from part 1, you'll see that it should be around 80% accuracy, about halfway between 74% and 88% accuracy."
+    "If you check out the accuracy of your frontend application again in the notebook from part 1, you'll see that it should be around 78-80% accuracy, about halfway between the accuracies of your SKLearn (68-74%) and TensorFlow (88%) accuracy models."
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: Typo: s/the accuracies of your SKLearn (68-74%) and TensorFlow (88%) accuracy models/the accuracies of your SKLearn (68-74%) and TensorFlow (88%) models

File: examples/tutorial/tutorial_part_two.ipynb [hunk 13]
```diff
@@ -310,15 +338,6 @@
     "\n",
     "*Head back to the notebook from part 1. When you're done watching the accuracy of your application, stop the cell (hit the little \"stop\" square in the notebook toolbar).*"
    ]
-  },
```
Comment (dcrankshaw) [inline examples/tutorial/tutorial_part_two.ipynb]: As discussed, put the Cleanup section here. Find a fun warning/caution image to put at the start of the cleanup section and make sure it's clear to run cleanup only when you are completely done with the tutorial.

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
In the diagrams, change "CIFAR data" to "CIFAR images." I think this will make it slightly clearer that this is a computer vision task and the data we are sending are images.

Also, can we visually diffentiate the `clipper` object doing management tasks with querying the application? Maybe make query the application a vertical flow (have the arrow from Python app to Docker be vertical)?

As discussed, move the entire Cleanup section to part 2.

**Comment by dcrankshaw** [comment]
This looks good! The diagrams are a nice addition.

**Review by dcrankshaw — APPROVED** [review]

**Comment by dcrankshaw** [comment]
jenkins test this please
